File: 45. Jump Game II.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        step = [-1] * (n)
        step[0] = 0
        
        for i, s in enumerate(nums):
            j = min(i + s, n - 1)
            if step[j] == -1 or step[i] + 1 < step[j]:
                step[j] = step[i] + 1
                for k in range(i+1, j):
                    if step[k] == -1 or step[i] + 1 < step[k]:
                        step[k] = step[i] + 1
        return step[-1]
            
        

File: 45. Jump Game II.go

func jump(nums []int) int {
	min := func(a, b int) int {
		if a < b {
			return a
		}
		return b
	}

	n := len(nums)

	dp := make([]int, n)
	for i := 1; i < n; i++ {
		dp[i] = n
	}

	for i := 0; i < n; i++ {
		for j := min(i+nums[i], n-1); j > i; j-- {
			if dp[j] > dp[i]+1 {
				dp[j] = dp[i] + 1
			} else {
				break
			}
		}
	}
	return dp[n-1]
}
